### ISSUE TITLE
Allow specification of non-builtin backend

### ISSUE THREAD

**Issue by macks22** [bambinos/bambi#201]
**Limitation**
I love the model building interface `bambi` provides, and I'd like to write a custom backend for a private project. As I understand it, the options for setting a custom backend are currently:

1. Subclass `bambi.Model` to override `_set_backend` to add support for my custom backend as another string. Then require users to use this subclass instead of `bambi.Model`.
2. Same as above but monkey-patch `bambi.Model` instead by setting a custom `_set_backend` method on import of my package.

I also considered trying to set the `_backend_name` private instance variable, but that won't work on its own because calls to `build` ultimately call `_set_backend`, which assumes it's a string in `{'pymc3', 'stan'}`, so we're back to one of the two options above.

Neither of these two options are particularly desirable. Monkey-patching is an anti-pattern in this case and requiring users to use a Model subclass works but seems unnecessary.

**Potential Solutions**
Some options for implementing this:

1.  Allow `backend to specified` as either a `str` (current behavior) or as `bambi.base.BackEnd` (new) in `_set_backend`, `build`, and `fit`. In the latter case, support setting of `_backend_name` by adding a class attribute to the `BackEnd` interface called name, which is a `str` that specifies the name. This seems like the simplest option to me.
2. Provide some kind of backend registration mechanism: this would be something like a plugin paradigm, where new backends register themselves using some kind of `bambi`-provided registry. One lightweight option for this would be to maintain a mapping on the `BackEnd` class itself and give it a metaclass that registers new subclasses by adding them to this mapping (using the new `name` class attribute as a key perhaps) on creation.

**Comment by macks22** [comment]
This might also provide a good pathway to spin out the `stan` backend into its own python package, which can be maintained (or not) separately.

**Comment by aloctavodia** [comment]
Hi @macks22 thanks for opening this issue I will come back to you with a more concrete answer in a couple of days, sorry for the delay.

**Comment by macks22** [comment]
Thanks @aloctavodia! Looking forward to hearing your thoughts on a preferred approach. I'd be to happy to put up a PR but want to make sure the design matches your preference.

**Comment by aloctavodia** [comment]
Sorry (again) for the delay. I think option 1 is preferred. Thanks in advance!

**Comment by macks22** [comment]
Great, thanks for the guidance! Will hopefully have time to get a PR up for this in the next few weeks.

**Comment by tomicapretto** [comment]
I don't think this will be possible now. It will require having other backends implemented as well as modifying/extending some of our classes which do use functions from the PyMC backend.
